File: packages/core/src/nostr/ingest.ts
/**
 * Inbound command ingest: Nostr event → NIP-44 decrypt → codec validation →
 * typed dispatch. Invalid payloads are logged and dropped — this layer NEVER
 * throws into the subscription callback.
 *
 * Ported from codedeck-bridge-vscode/src/nostrRelay.ts:
 * - processedEventIds LRU dedup (cap 1000) — relay reconnections with
 *   overlapping `since` windows replay events.
 * - 300s staleness cutoff (relays that don't enforce `since`).
 * - `since` selection: persisted last-seen − 5s grace, else a 5-minute window.
 * - The authorless pairing-window filter: the only path by which a
 *   not-yet-paired phone can reach the bridge, open only while pairing.
 * Redesigned: JSON.parse+cast replaced by decodePhoneToBridge (zod), and the
 * giant switch now dispatches to an all-optional CommandHandlers interface.
 */
import type { Filter } from 'nostr-tools/filter';
import type { NostrEvent } from 'nostr-tools/core';
import {
  COMMAND_KIND,
  decodePhoneToBridge,
  type PhoneToBridgeMessage,
} from '@codedeck/protocol';
import { decryptFrom } from './crypto';

// --- Filter builders ---

/** Main commands subscription: stored commands from paired phones only. */
export function buildCommandsFilter(opts: {
  bridgePubkey: string;
  phonePubkeys: readonly string[];
  since: number;
}): Filter {
  return {
    kinds: [COMMAND_KIND],
    '#p': [opts.bridgePubkey],
    authors: [...opts.phonePubkeys],
    since: opts.since,
  };
}

/** `since` for a (re)connect: persisted last-seen timestamp with a 5s grace to
 *  bridge crash gaps, falling back to a 5-minute window (ported). */
export function sinceForConnect(
  lastSeenTimestamp: number,
  nowSec: number = Math.floor(Date.now() / 1000),
): number {
  return lastSeenTimestamp > 0 ? lastSeenTimestamp - 5 : nowSec - 300;
}

/**
 * Pairing-window subscription: intentionally NO `authors` filter — that's the
 * whole point. An as-yet-unpaired phone can only reach the bridge while this
 * time-boxed window is open; a one-time token (checked by the pair-request
 * handler) gates acceptance.
 */
export function buildPairingFilter(opts: {
  bridgePubkey: string;
  nowSec?: number;
}): Filter {
  const now = opts.nowSec ?? Math.floor(Date.now() / 1000);
  return {
    kinds: [COMMAND_KIND],
    '#p': [opts.bridgePubkey],
    since: now - 5,
  };
}

// --- Typed dispatch ---

type Msg<K extends PhoneToBridgeMessage['type']> = Extract<PhoneToBridgeMessage, { type: K }>;

type Handler<K extends PhoneToBridgeMessage['type']> = (
  msg: Msg<K>,
  phonePubkeyHex: string,
) => void | Promise<void>;

/** One method per phone→bridge message type, all optional. Handlers may be
 *  async; rejections are caught and logged, never propagated. */
export interface CommandHandlers {
  onInput?: Handler<'input'>;
  onQuestionInput?: Handler<'question-input'>;
  onPermissionResponse?: Handler<'permission-res'>;
  onKeypress?: Handler<'keypress'>;
  onModeChange?: Handler<'mode'>;
  onEffortChange?: Handler<'effort'>;
  onModelChange?: Handler<'model'>;
  onSyncRequest?: Handler<'sync-request'>;
  onSyncAck?: Handler<'sync-ack'>;
  onCreateSession?: Handler<'create-session'>;
  onRefreshSessions?: Handler<'refresh-sessions'>;
  onCloseSession?: Handler<'close-session'>;
  onInterrupt?: Handler<'interrupt'>;
  onCreateFolder?: Handler<'create-folder'>;
  onUploadImage?: Handler<'upload-image'>;
  onUsageRequest?: Handler<'usage-request'>;
  onGsdRequest?: Handler<'gsd-request'>;
  onModelsRequest?: Handler<'models-request'>;
  onSetCredentials?: Handler<'set-credentials'>;
  onSetProviderProfile?: Handler<'set-provider-profile'>;
  onProviderProfilesRequest?: Handler<'provider-profiles-request'>;
  onSetDeviceConfig?: Handler<'set-device-config'>;
  onPairRequest?: Handler<'pair-request'>;
}

export interface CommandIngestOptions {
  secretKey: Uint8Array;
  handlers: CommandHandlers;
  /**
   * CDX-050: called for EVERY validly-decoded main-path command with the
   * sender's advertised capability strings (`caps`, `[]` when the command
   * carries none — pre-CDX-050 phones omit the field). Feeds the bridge's
   * phone-capability registry, which gates diff-entry emission.
   */
  onPhoneCaps?: (phonePubkeyHex: string, caps: readonly string[]) => void;
  /** When provided, main-path events whose author fails this check are dropped
   *  (the pairing path never consults it — unpaired phones are its point). */
  isPairedPhone?: (pubkeyHex: string) => boolean;
  log?: (msg: string) => void;
  /** Clock in ms — injectable for tests. */
  now?: () => number;
  /** Persisted cursor from a previous run (feeds sinceForConnect). */
  lastSeenTimestamp?: number;
  /**
   * Persisted dedup ids from a previous run. A restarted bridge resubscribes
   * with `since = lastSeen − 5s`, so already-processed commands inside that
   * grace window are REPLAYED by the relay — without seeding the dedup set they
   * would be re-executed (duplicate create-session, re-sent inputs). Found by
   * the CDX-008 restart contract test.
   */
  processedEventIds?: readonly string[];
}

export class CommandIngest {
  private readonly secretKey: Uint8Array;
  private readonly handlers: CommandHandlers;
  private readonly onPhoneCaps?: (phonePubkeyHex: string, caps: readonly string[]) => void;
  private readonly isPairedPhone?: (pubkeyHex: string) => boolean;
  private readonly logFn?: (msg: string) => void;
  private readonly now: () => number;

  // --- Event deduplication ---
  // Recently processed nostr event IDs, insertion-ordered; oldest evicted at
  // the cap. Prevents replayed events (overlapping `since` windows) from
  // triggering duplicate side effects like spawning multiple sessions.
  private readonly processedEventIds = new Set<string>();
  private static readonly MAX_PROCESSED_EVENT_IDS = 1000;

  /** Safety net: ignore events this much older than now, in case relays don't
   *  enforce `since` (ported). */
  private static readonly MAX_EVENT_AGE_SECONDS = 300;

  // Most recent event timestamp we processed — persisted by the host and used
  // as `since` on reconnect to bridge crash gaps.
  private _lastSeenTimestamp: number;

  constructor(options: CommandIngestOptions) {
    this.secretKey = options.secretKey;
    this.handlers = options.handlers;
    this.onPhoneCaps = options.onPhoneCaps;
    this.isPairedPhone = options.isPairedPhone;
    this.logFn = options.log;
    this.now = options.now ?? Date.now;
    this._lastSeenTimestamp = options.lastSeenTimestamp ?? 0;
    for (const id of (options.processedEventIds ?? []).slice(-CommandIngest.MAX_PROCESSED_EVENT_IDS)) {
      this.processedEventIds.add(id);
    }
  }

  /** Timestamp of the last processed event (for persistence). */
  get lastSeenTimestamp(): number {
    return this._lastSeenTimestamp;
  }

  /** Snapshot of the dedup ids, oldest first (for persistence alongside the cursor). */
  processedIds(): string[] {
    return [...this.processedEventIds];
  }

  /** Handle an event from the main (paired-authors) subscription. Never throws. */
  handleEvent(event: NostrEvent): void {
    const nowSec = Math.floor(this.now() / 1000);
    if (event.created_at < nowSec - CommandIngest.MAX_EVENT_AGE_SECONDS) {
      this.log(`[Ingest] Ignoring stale event (${nowSec - event.created_at}s old)`);
      return;
    }

    // CDX-013: authorization BEFORE dedup bookkeeping — events from unknown
    // pubkeys must not consume LRU slots (a flood could evict legitimate ids
    // and open a replay window).
    if (this.isPairedPhone && !this.isPairedPhone(event.pubkey)) {
      this.log(`[Ingest] Ignoring event from unknown pubkey: ${event.pubkey.slice(0, 8)}...`);
      return;
    }

    if (!this.markProcessed(event.id)) { return; } // duplicate

    let plaintext: string;
    try {
      plaintext = decryptFrom(this.secretKey, event.pubkey, event.content);
    } catch (err) {
      this.log(`[Ingest] Failed to decrypt event ${event.id.slice(0, 8)}...: ${err}`);
      return;
    }

    const decoded = decodePhoneToBridge(plaintext);
    if (!decoded.ok) {
      this.log(`[Ingest] Dropping invalid payload from ${event.pubkey.slice(0, 8)}...: ${decoded.error}`);
      return;
    }

    // Track last-seen event timestamp for the crash-recovery since filter.
    if (event.created_at > this._lastSeenTimestamp) {
      this._lastSeenTimestamp = event.created_at;
    }

    this.log(`[Ingest] Received ${decoded.msg.type} from ${event.pubkey.slice(0, 8)}...`);
    // CDX-050: record the sender's advertised capabilities on every valid
    // command — a phone that omits `caps` is recorded as [] (pre-CDX-050),
    // which keeps diff-entry emission off for mixed fleets.
    try {
      this.onPhoneCaps?.(event.pubkey, decoded.msg.caps ?? []);
    } catch (err) {
      this.log(`[Ingest] onPhoneCaps error: ${err}`);
    }
    this.dispatch(decoded.msg, event.pubkey);
  }

  /**
   * Handle an event from the authorless pairing-window subscription. The open
   * filter attracts junk we can't decrypt — dropped silently. Only
   * `pair-request` is dispatched from this path; token validation is the
   * onPairRequest handler's job.
   */
  handlePairingEvent(event: NostrEvent): void {
    // CDX-013: same staleness cutoff as the main path. The relay-side `since`
    // filter usually enforces this, but a hostile/buggy relay could replay old
    // events — the token check downstream would still refuse them; this keeps
    // them out of the decrypt + dedup machinery entirely.
    const nowSec = Math.floor(this.now() / 1000);
    if (event.created_at < nowSec - CommandIngest.MAX_EVENT_AGE_SECONDS) {
      return;
    }

    let plaintext: string;
    try {
      plaintext = decryptFrom(this.secretKey, event.pubkey, event.content);
    } catch {
      return; // not for us / not decryptable — ignore silently
    }

    // Dedup AFTER decrypt: junk on the authorless window must not consume LRU
    // slots shared with the main command path (CDX-013).
    if (!this.markProcessed(event.id)) { return; } // duplicate

    const decoded = decodePhoneToBridge(plaintext);
    if (!decoded.ok) { return; }
    if (decoded.msg.type !== 'pair-request') { return; }

    this.log(`[Ingest] Valid pair-request from "${decoded.msg.label}" (${event.pubkey.slice(0, 8)}...)`);
    this.invoke('onPairRequest', this.handlers.onPairRequest, decoded.msg, event.pubkey);
  }

  /** Record an event id in the dedup LRU. Returns false when already seen. */
  private markProcessed(id: string): boolean {
    if (this.processedEventIds.has(id)) { return false; }
    this.processedEventIds.add(id);
    if (this.processedEventIds.size > CommandIngest.MAX_PROCESSED_EVENT_IDS) {
      const first = this.processedEventIds.values().next().value;
      if (first !== undefined) { this.processedEventIds.delete(first); }
    }
    return true;
  }

  /** Exhaustive over the union — an unrouted message type is a compile error. */
  private dispatch(msg: PhoneToBridgeMessage, pubkey: string): void {
    switch (msg.type) {
      case 'input': return this.invoke('onInput', this.handlers.onInput, msg, pubkey);
      case 'question-input': return this.invoke('onQuestionInput', this.handlers.onQuestionInput, msg, pubkey);
      case 'permission-res': return this.invoke('onPermissionResponse', this.handlers.onPermissionResponse, msg, pubkey);
      case 'keypress': return this.invoke('onKeypress', this.handlers.onKeypress, msg, pubkey);
      case 'mode': return this.invoke('onModeChange', this.handlers.onModeChange, msg, pubkey);
      case 'effort': return this.invoke('onEffortChange', this.handlers.onEffortChange, msg, pubkey);
      case 'model': return this.invoke('onModelChange', this.handlers.onModelChange, msg, pubkey);
      case 'sync-request': return this.invoke('onSyncRequest', this.handlers.onSyncRequest, msg, pubkey);
      case 'sync-ack': return this.invoke('onSyncAck', this.handlers.onSyncAck, msg, pubkey);
      case 'create-session': return this.invoke('onCreateSession', this.handlers.onCreateSession, msg, pubkey);
      case 'refresh-sessions': return this.invoke('onRefreshSessions', this.handlers.onRefreshSessions, msg, pubkey);
      case 'close-session': return this.invoke('onCloseSession', this.handlers.onCloseSession, msg, pubkey);
      case 'interrupt': return this.invoke('onInterrupt', this.handlers.onInterrupt, msg, pubkey);
      case 'create-folder': return this.invoke('onCreateFolder', this.handlers.onCreateFolder, msg, pubkey);
      case 'upload-image': return this.invoke('onUploadImage', this.handlers.onUploadImage, msg, pubkey);
      case 'usage-request': return this.invoke('onUsageRequest', this.handlers.onUsageRequest, msg, pubkey);
      case 'gsd-request': return this.invoke('onGsdRequest', this.handlers.onGsdRequest, msg, pubkey);
      case 'models-request': return this.invoke('onModelsRequest', this.handlers.onModelsRequest, msg, pubkey);
      case 'set-credentials': return this.invoke('onSetCredentials', this.handlers.onSetCredentials, msg, pubkey);
      case 'set-provider-profile': return this.invoke('onSetProviderProfile', this.handlers.onSetProviderProfile, msg, pubkey);
      case 'provider-profiles-request': return this.invoke('onProviderProfilesRequest', this.handlers.onProviderProfilesRequest, msg, pubkey);
      case 'set-device-config': return this.invoke('onSetDeviceConfig', this.handlers.onSetDeviceConfig, msg, pubkey);
      case 'pair-request': return this.invoke('onPairRequest', this.handlers.onPairRequest, msg, pubkey);
      default: {
        const exhaustive: never = msg;
        this.log(`[Ingest] Ignoring unhandled message type: ${(exhaustive as { type: string }).type}`);
      }
    }
  }

  /** Call a handler, containing both sync throws and async rejections. */
  private invoke<M>(
    name: string,
    handler: ((msg: M, phonePubkeyHex: string) => void | Promise<void>) | undefined,
    msg: M,
    pubkey: string,
  ): void {
    if (!handler) { return; }
    try {
      Promise.resolve(handler(msg, pubkey)).catch((err) => {
        this.log(`[Ingest] ${name} handler error: ${err}`);
      });
    } catch (err) {
      this.log(`[Ingest] ${name} handler error: ${err}`);
    }
  }

  private log(msg: string): void {
    this.logFn?.(msg);
  }
}

File: crates/bridge-core/src/ingest.rs
//! Inbound commands: a relay event becomes a decoded phone message, or is
//! dropped (with a log line) — never an error for the caller.
//!
//! - Events older than [`MAX_EVENT_AGE_SECS`] are dropped: a relay that does
//!   not honour `since` must not replay old commands.
//! - On the standing subscription, only paired phones are heard, and the
//!   check comes before dedup bookkeeping so a flood from strangers cannot
//!   evict real ids from the dedup set and reopen a replay window.
//! - Event ids are remembered (the last [`MAX_PROCESSED_IDS`]) and persisted:
//!   a reconnect subscribes with `since` = last seen − 5 s, so the relay
//!   replays recent commands, and without the ids they would run twice
//!   (duplicate sessions, re-sent input).
//! - The pairing window's subscription has no author filter, so it attracts
//!   junk: anything that does not decrypt is dropped silently and only then
//!   takes a dedup slot, and only a `pair-request` gets through.

use std::collections::{HashSet, VecDeque};

use protocol::commands::{PairRequestMsg, PhoneToBridge};
use protocol::crypto::{decrypt_from, Keypair};

use crate::io::InboundEvent;

pub const MAX_PROCESSED_IDS: usize = 1000;
pub const MAX_EVENT_AGE_SECS: u64 = 300;

/// Where a (re)connect's command subscription starts: the last event seen
/// minus a 5 s grace for a crash gap, else the last five minutes.
pub fn since_for_connect(last_seen: u64, now_secs: u64) -> u64 {
    if last_seen > 0 {
        last_seen.saturating_sub(5)
    } else {
        now_secs.saturating_sub(300)
    }
}

/// The standing subscription: command events tagged to the bridge from
/// these authors, created at or after `since`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CommandsFilter {
    pub authors: Vec<String>,
    pub since: u64,
}

pub struct Ingest {
    order: VecDeque<String>,
    seen: HashSet<String>,
    last_seen: u64,
}

fn short(pubkey: &str) -> &str {
    pubkey.get(..8).unwrap_or(pubkey)
}

impl Ingest {
    /// Resume from a persisted cursor and dedup set (oldest id first).
    pub fn new(last_seen: u64, processed: Vec<String>) -> Self {
        let mut ingest = Self { order: VecDeque::new(), seen: HashSet::new(), last_seen };
        let skip = processed.len().saturating_sub(MAX_PROCESSED_IDS);
        for id in processed.into_iter().skip(skip) {
            ingest.mark(&id);
        }
        ingest
    }

    /// Created-at of the newest command handled (seconds).
    pub fn last_seen(&self) -> u64 {
        self.last_seen
    }

    /// The dedup ids, oldest first, for persisting beside the cursor.
    pub fn processed_ids(&self) -> Vec<String> {
        self.order.iter().cloned().collect()
    }

    /// An event from the standing subscription.
    pub fn accept(
        &mut self,
        event: &InboundEvent,
        keys: &Keypair,
        now_secs: u64,
        is_paired: impl Fn(&str) -> bool,
    ) -> Option<PhoneToBridge> {
        if event.created_at + MAX_EVENT_AGE_SECS < now_secs {
            log::info!("[Ingest] Ignoring stale event ({}s old)", now_secs - event.created_at);
            return None;
        }
        if !is_paired(&event.pubkey) {
            log::info!("[Ingest] Ignoring event from unknown pubkey: {}...", short(&event.pubkey));
            return None;
        }
        if !self.mark(&event.id) {
            return None;
        }
        let plaintext = match decrypt_from(&keys.secret_key, &event.pubkey, &event.content) {
            Ok(p) => p,
            Err(err) => {
                log::info!("[Ingest] Failed to decrypt event {}...: {err}", short(&event.id));
                return None;
            }
        };
        let msg = match protocol::decode_phone_to_bridge(&plaintext) {
            Ok(msg) => msg,
            Err(err) => {
                log::info!("[Ingest] Dropping invalid payload from {}...: {err}", short(&event.pubkey));
                return None;
            }
        };
        self.last_seen = self.last_seen.max(event.created_at);
        Some(msg)
    }

    /// An event from the pairing window's subscription.
    pub fn accept_pairing(&mut self, event: &InboundEvent, keys: &Keypair, now_secs: u64) -> Option<PairRequestMsg> {
        if event.created_at + MAX_EVENT_AGE_SECS < now_secs {
            return None;
        }
        let plaintext = decrypt_from(&keys.secret_key, &event.pubkey, &event.content).ok()?;
        if !self.mark(&event.id) {
            return None;
        }
        match protocol::decode_phone_to_bridge(&plaintext) {
            Ok(PhoneToBridge::PairRequest(msg)) => {
                // The label is unverified sender input (the token is checked
                // later): Debug-quoted, so it cannot inject log lines.
                log::info!("[Ingest] Valid pair-request from {:?} ({}...)", msg.label, short(&event.pubkey));
                Some(msg)
            }
            _ => None,
        }
    }

    /// Remember an id; false when it was already there.
    fn mark(&mut self, id: &str) -> bool {
        if !self.seen.insert(id.to_string()) {
            return false;
        }
        self.order.push_back(id.to_string());
        if self.order.len() > MAX_PROCESSED_IDS {
            if let Some(old) = self.order.pop_front() {
                self.seen.remove(&old);
            }
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use protocol::crypto::{encrypt_to, generate_keypair};

    struct Fixture {
        bridge: Keypair,
        phone: Keypair,
        ingest: Ingest,
        n: u32,
    }

    const NOW: u64 = 1_800_000_000;

    impl Fixture {
        fn new() -> Self {
            Self { bridge: generate_keypair(), phone: generate_keypair(), ingest: Ingest::new(0, vec![]), n: 0 }
        }
        fn event_from(&mut self, from: &Keypair, json: &str, created_at: u64) -> InboundEvent {
            self.n += 1;
            InboundEvent {
                id: format!("ev{}", self.n),
                pubkey: from.pubkey_hex.clone(),
                created_at,
                content: encrypt_to(&from.secret_key, &self.bridge.pubkey_hex, json).unwrap(),
            }
        }
        fn event(&mut self, json: &str) -> InboundEvent {
            let phone = self.phone.clone();
            self.event_from(&phone, json, NOW)
        }
        fn accept(&mut self, event: &InboundEvent) -> Option<PhoneToBridge> {
            let phone = self.phone.pubkey_hex.clone();
            self.ingest.accept(event, &self.bridge, NOW, |pk| pk == phone)
        }
    }

    const INPUT: &str = r#"{"type":"input","sessionId":"s","text":"hi"}"#;

    #[test]
    fn a_valid_command_decrypts_decodes_and_advances_the_cursor() {
        let mut f = Fixture::new();
        let ev = f.event(INPUT);
        assert!(matches!(f.accept(&ev), Some(PhoneToBridge::Input(m)) if m.text == "hi"));
        assert_eq!(f.ingest.last_seen(), NOW);
    }

    #[test]
    fn malformed_unknown_and_undecryptable_payloads_are_dropped() {
        let mut f = Fixture::new();
        let bad_json = f.event("{not json");
        let unknown = f.event(r#"{"type":"teleport"}"#);
        let mut garbage = f.event(INPUT);
        garbage.content = "AAAA".into();
        assert!(f.accept(&bad_json).is_none());
        assert!(f.accept(&unknown).is_none());
        assert!(f.accept(&garbage).is_none());
        assert_eq!(f.ingest.last_seen(), 0, "nothing handled, cursor unmoved");
    }

    #[test]
    fn events_older_than_five_minutes_are_dropped() {
        let mut f = Fixture::new();
        let phone = f.phone.clone();
        let old = f.event_from(&phone, INPUT, NOW - 301);
        let edge = f.event_from(&phone, INPUT, NOW - 300);
        assert!(f.accept(&old).is_none());
        assert!(f.accept(&edge).is_some());
    }

    #[test]
    fn a_replayed_event_id_runs_once() {
        let mut f = Fixture::new();
        let ev = f.event(INPUT);
        assert!(f.accept(&ev).is_some());
        assert!(f.accept(&ev).is_none());
    }

    #[test]
    fn the_dedup_set_is_an_lru_capped_at_1000() {
        let mut f = Fixture::new();
        let first = f.event(INPUT);
        assert!(f.accept(&first).is_some());
        for i in 0..MAX_PROCESSED_IDS {
            assert!(f.ingest.mark(&format!("filler{i}")));
        }
        assert!(f.accept(&first).is_some(), "the oldest id aged out");
        assert_eq!(f.ingest.processed_ids().len(), MAX_PROCESSED_IDS);
    }

    #[test]
    fn a_persisted_dedup_set_survives_a_restart() {
        let mut f = Fixture::new();
        let ev = f.event(INPUT);
        assert!(f.accept(&ev).is_some());
        f.ingest = Ingest::new(f.ingest.last_seen(), f.ingest.processed_ids());
        assert!(f.accept(&ev).is_none(), "the relay's grace-window replay is a no-op");
    }

    #[test]
    fn strangers_are_dropped_before_they_take_a_dedup_slot() {
        let mut f = Fixture::new();
        let stranger = generate_keypair();
        let ev = f.event_from(&stranger, INPUT, NOW);
        assert!(f.accept(&ev).is_none());
        assert!(f.ingest.processed_ids().is_empty());
    }

    #[test]
    fn the_pairing_path_hears_strangers_but_only_their_pair_requests() {
        let mut f = Fixture::new();
        let stranger = generate_keypair();
        let pair = r#"{"type":"pair-request","npub":"n","pubkeyHex":"aa","label":"Pixel","token":"t"}"#;
        let ev = f.event_from(&stranger, pair, NOW);
        let req = f.ingest.accept_pairing(&ev, &f.bridge, NOW).unwrap();
        assert_eq!(req.label, "Pixel");
        assert!(f.ingest.accept_pairing(&ev, &f.bridge, NOW).is_none(), "deduped");

        let input = f.event_from(&stranger, INPUT, NOW);
        assert!(f.ingest.accept_pairing(&input, &f.bridge, NOW).is_none());
        let stale = f.event_from(&stranger, pair, NOW - 301);
        assert!(f.ingest.accept_pairing(&stale, &f.bridge, NOW).is_none());
    }

    #[test]
    fn junk_on_the_pairing_window_takes_no_dedup_slot() {
        let mut f = Fixture::new();
        let stranger = generate_keypair();
        let mut junk = f.event_from(&stranger, INPUT, NOW);
        junk.content = "not ciphertext".into();
        assert!(f.ingest.accept_pairing(&junk, &f.bridge, NOW).is_none());
        assert!(f.ingest.processed_ids().is_empty());
    }

    #[test]
    fn since_for_connect_uses_the_cursor_with_a_grace_else_five_minutes() {
        assert_eq!(since_for_connect(1000, 5000), 995);
        assert_eq!(since_for_connect(0, 5000), 4700);
    }
}
